### PR TITLE
Fix reading numpy arrays on BE arch.

### PR DESCRIPTION
**Pull request by sebpuetz**

### CODE REVIEW

**Comment by sebpuetz** [comment]
The failed job is due to travis messing up. Failed to download the Python archive...

Not sure whether it makes sense to store arrays as LE on BE systems. It's also possible to `byteswap()` the in-memory array for BE architectures after reading to keep it in the native order. But that won't work with the memmap array since `byteswap()` would either alter the on-disk file or create a byte-swapped copy in-memory. 

Copying the `finalfusion-rust` approach and swapping the bytes in the returned embeddings doesn't seem feasible either.

We'd have to track when the backing array is on-disk with a class attribute since the `np.memmap` type gets lost somewhere:

```Python
>>> array = np.memmap(file)
>>> type(array).mro()
[numpy.memmap, numpy.ndarray, object]

>>> storage = NdArray(array)
>>> type(storage).mro()
[finalfusion.storage.ndarray.NdArray,
 numpy.ndarray,
 finalfusion.storage.storage.Storage,
 finalfusion.io.Chunk,
 abc.ABC,
 object]
```

The `__getitem__` implementation would need to consider the `mmap` attribute and also check for the endianness of the system.

This would solve embedding retrieval but doing any other kinds of numpy operations on a memory mapped matrix on BE architectures would still yield incorrect results since reads would still be incorrect.

So I see a few options:
* disable memapped arrays on BE systems
* keep memmapped arrays as LE on BE systems
* ensure that **all** arrays are LE, no matter the architecture

I'm currently leaning towards disabling memmap support on BE since I don't see a clean workaround. Even if we somehow track whether we just memmapped an array, there's no clean solution that would both return the embeddings as BE and correctly handle other numpy operations.

**Comment by danieldk** [comment]
> Copying the `finalfusion-rust` approach and swapping the bytes in the returned embeddings doesn't seem feasible either.

But IIRC we don't do that in `finalfusion-rust`. We just convert for the native endianness when reading in the embeddings and do not support memory-mapped embeddings on BE platforms (yet). Memory mapping would be possible by swapping on the fly (as you suggested), but that does not work for e.g. similarity queries if you want to do a fast matrix-vector multiplication.

> So I see a few options:
>     * disable memapped arrays on BE systems

Since this is what we currently in `finalfusion-rust`, this seems to be the most consistent.

I think in the future we could add an alternative that relies on seeks + reads. It wouldn't be able to offer the `StorageView` interface, but would work for retrieving embeddings, while keeping memory usage low like memory mapping.

**Comment by sebpuetz** [comment]
> > Copying the `finalfusion-rust` approach and swapping the bytes in the returned embeddings doesn't seem feasible either.
> 
> But IIRC we don't do that in `finalfusion-rust`. We just convert for the native endianness when reading in the embeddings and do not support memory-mapped embeddings on BE platforms (yet). Memory mapping would be possible by swapping on the fly (as you suggested), but that does not work for e.g. similarity queries if you want to do a fast matrix-vector multiplication.

```Rust
        #[cfg(target_endian = "big")]
        LittleEndian::from_slice_f32(
            embedding
                .as_slice_mut()
                .expect("Cannot borrow vector as mutable slice"),
        );
```
from `finalfusion/src/chunks/storage/array.rs`, so we mmap the array as-is and swap bytes when retrieving embeddings, we only disable writing and `StorageView` for `MmapArray`.

> > So I see a few options:
> > * disable memapped arrays on BE systems
> 
> Since this is what we currently in `finalfusion-rust`, this seems to be the most consistent.

Even with the point above I'm still in favor of just disabling mmap in general for BE systems.

**Comment by danieldk** [comment]
> from `finalfusion/src/chunks/storage/array.rs`, so we mmap the array as-is and swap bytes when retrieving embeddings, we only disable writing and `StorageView` for `MmapArray`.

Oh, I completely forgot about that :open_mouth: . The other day I was helping Patricia and we found that dpar still used the `rust2vec` crate and it turns out that it also supported the FiFu format already. In my head `rust2vec` never did that. It's interesting how lossy the :brain: is. At least mine :wink:.

> Even with the point above I'm still in favor of just disabling mmap in general for BE systems.

I think it is nice to have BE support, but I don't think we have to support every feature, there are not that many BE systems around. So, I think this is a reasonable compromise.

File: src/finalfusion/io.py [hunk 1]
```diff
@@ -307,3 +310,17 @@ def _read_required_chunk_header(file: BinaryIO) -> Tuple[ChunkIdentifier, int]:
     if val is None:
         raise FinalfusionFormatError('could not read chunk header.')
     return val
+
+
+def _serialize_array_as_le(file: BinaryIO, array: np.ndarray):
+    native_is_le = sys.byteorder == "little"
+    array_bo = array.dtype.byteorder
+    array_is_le = array_bo in "<|" or array_bo == "=" and native_is_le
```
Comment (danieldk) [inline src/finalfusion/io.py]: Add parentheses to make precedence explicit.

File: tests/test_storage.py [hunk 2]
```diff
@@ -22,6 +22,8 @@ def test_read_array(tests_root, vocab_array_tuple):
 
 
 def test_mmap_array(tests_root, vocab_array_tuple):
+    if sys.byteorder == "big":
```
Comment (danieldk) [inline tests/test_storage.py]: Use the `skipif` decorator for these cases:

https://docs.pytest.org/en/latest/skipping.html#id1

**Comment by sebpuetz** [comment]
> > from `finalfusion/src/chunks/storage/array.rs`, so we mmap the array as-is and swap bytes when retrieving embeddings, we only disable writing and `StorageView` for `MmapArray`.
> 
> Oh, I completely forgot about that open_mouth . The other day I was helping Patricia and we found that dpar still used the `rust2vec` crate and it turns out that it also supported the FiFu format already. In my head `rust2vec` never did that. It's interesting how lossy the :brain: is. At least mine :wink:.

I think we started development of the finalfusion format in `rust2vec`!

> > Even with the point above I'm still in favor of just disabling mmap in general for BE systems.
> 
> I think it is nice to have BE support, but I don't think we have to support every feature, there are not that many BE systems around. So, I think this is a reasonable compromise.

Sure, is this one good to go then?

**Review by danieldk — CHANGES_REQUESTED** [review]

**Review by danieldk — APPROVED** [review]

**Pull request merged**